Added property to find the number of ticks of the slider automatically.

phnix-dev opened the pull request:
I found it odd that we must set the number of ticks on a slider by ourselves, I assume we want the grabber to be aligned with the ticks.
Since we have to find by ourselves the correct number of ticks, it becomes very tedious and difficult to align them with steps.

So, I added a bool in the slider class to make sure that if it is true, it will find the right number of ticks for the number of steps and if it is impossible to align the steps on the grabber, it will set the steps to an appropriate lower number of steps to align the ticks on the grabber.

The change is compatible with `3.x`

phnix-dev committed:
added automatic a step adjuster

To align the grabber with the ticks.

Calinou commented:
How does this handle very small step values (and large slider ranges), where hundreds of ticks would have to be drawn? Should there be a minimum separation threshold or a maximum number of ticks this feature can draw (e.g. 50)?

phnix-dev committed:
fixed the maximum possible value of the step

The maximum step cannot be greater than the distance from min to max.
(50 to 100 => max step of 50)

MrPhnix commented:
First, I would like to explain a little better what my code does.

I assume that the user wants the number of ticks to be found automatically when choosing the step.

To do this I divide the distance (max - min) by the steps, which gives me the number of ticks.
If the result does not give a remainder of 0 then the grabber will not be aligned with the ticks, because you cannot have a decimal number of ticks. Therefore, I reduce the step to a lower value.


For example, if the user has a slider from 0 to 1000 and wants a step of 1 then there will be 1001 ticks.
I think we should not put a maximum number of ticks to this feature. I prefer to let the user realize, for this example, that a step of 1 is a too small value.

phnix-dev committed:
fixed the signature of the "is_adjusting_step" function

phnix-dev committed:
added "adjust_step" of the slider to the docs

phnix-dev committed:
formatted code with clang-format

phnix-dev committed:
Renamed slider property

phnix-dev committed:
replace update by queue_redraw

phnix-dev committed:
update slider docs

MrPhnix commented:
I renamed the property to find_ticks, which makes more sense than ajust_step.

YuriSizov commented on scene/gui/slider.cpp:
```diff
@@ -189,6 +189,35 @@ void Slider::_notification(int p_what) {
+				float step = get_step(), dist = get_max() - get_min();
+				int idist = dist;
+				if (step > dist) {
+					ticks = 0;
```
I don't think it should directly modify the exposed property. There can be an internal structure or set of properties that are affected by both the fixed value and the computed value.

And it certainly shouldn't be done on a each draw call. This value is static as long as the number of steps doesn't change. So it should be computed once, when the relevant properties change, and only used during the drawing process.

YuriSizov commented on scene/gui/slider.cpp:
```diff
@@ -325,6 +370,7 @@ void Slider::_bind_methods() {
 	ADD_PROPERTY(PropertyInfo(Variant::BOOL, "scrollable"), "set_scrollable", "is_scrollable");
 	ADD_PROPERTY(PropertyInfo(Variant::INT, "tick_count", PROPERTY_HINT_RANGE, "0,4096,1"), "set_ticks", "get_ticks");
 	ADD_PROPERTY(PropertyInfo(Variant::BOOL, "ticks_on_borders"), "set_ticks_on_borders", "get_ticks_on_borders");
+	ADD_PROPERTY(PropertyInfo(Variant::BOOL, "find_ticks"), "set_find_ticks", "is_finding_ticks");
```
I think that the name is not clear, and neither is the description for that matter. It could be something like `ticks_from_steps`. Setting this value should also probably make the property for the fixed number of ticks readonly (you would need to use `_validate_property`).

YuriSizov commented on scene/gui/slider.cpp:
```diff
@@ -189,6 +189,35 @@ void Slider::_notification(int p_what) {
+						factors.sort();
+						for (int i = factors.size() - 1; i >= 0; i--) {
+							if (step >= factors[i]) {
+								set_step(factors[i]);
```
It shouldn't change the number of steps. This becomes too convoluted if we have properties indirectly affect each other like that.

I also don't see this being useful in principle, because you want your users to be able to adjust the values with the precision specified in steps, physical limitations should be handled differently.

YuriSizov reviewed (changes requested):
I have some notes about the implementation, but it also makes me think that this is missing a proper proposal. Because the intended behavior here is very specific and more complex than the title leads to believe.

Aligning the number of ticks to the number of steps makes sense. Changing the number of steps based on the visual constrains doesn't, IMO. But it would also make sense to display ticks every N steps, with some sort of ratio. A proposal would be able to elaborate on this and provide a more complete solution, because this so far feels very hacky and very specific.

MrPhnix commented:
Thank you very much for the feedback. 
You are right, it's an overly complicated way to line up the ticks on the steps.

I have also been thinking more about this issue, and I think I have found the underlying problem:
The range is not the problem, it is the tick itself since the only way to place a tick is to enter a value to divide the range, but the range uses steps and not subdivisions*.

It is the placement of the ticks that need to be changed, to be directly aligned with the steps.
I don't think the choice of the number of ticks is important, since we just want them to be aligned with the steps.

I will try to make a proposal to talk more about it.
Thanks 😀

<br>
* Explanation of what the subdivision means to me. (How the ticks are placed in relation to how the step works)

![explanations](https://user-images.githubusercontent.com/76911907/209355159-dfea00da-d045-42c2-a325-c909482ae2a3.png)
